File: scripts/discover_overlaps.py
"""Discover conceptual overlaps between the article and the Kindle book.

Uses the hybrid search engine to find passages in each document that
are semantically similar to passages from the other, then creates
edges recording the thematic connections.

Documents:
  - Article: "A Biblical Case Against Government Economic Redistribution"
  - Book: "The Ethics of Money Production" by Jörg Guido Hülsmann
"""

from __future__ import annotations

import asyncio
import sys
from uuid import UUID

import structlog

from research_engine.adapters.storage.postgres.engine import transaction
from research_engine.composition import build_container
from research_engine.config import load_settings
from research_engine.domain.common import NodeKind
from research_engine.domain.edges import EdgeDraft
from research_engine.domain.passages import SearchFilters, SearchQuery

logger = structlog.get_logger()

# Document IDs from the database
ARTICLE_DOC_ID = UUID("019dbaea-e844-7201-83fe-69200d255ca5")
BOOK_DOC_ID = UUID("019dbac5-b33c-74e2-9a25-2117d837dcda")

# The 8 conceptual themes identified in the analysis, with search queries
# that capture each theme from both perspectives.
THEMES = [
    {
        "id": "property_rights_as_moral_foundation",
        "label": "Property Rights as a Moral Foundation",
        "queries": [
            "property rights moral foundation sacred inviolable justice",
            "taking property without authorization is theft",
            "private property sacred commutative justice",
        ],
    },
    {
        "id": "government_coercion",
        "label": "Government Force / Coercion as the Mechanism of the State",
        "queries": [
            "government monopoly on legitimate use of force compel compliance",
            "compulsion coercion by the ruler government force",
            "forcing citizens to use money of the government's choice",
        ],
    },
    {
        "id": "inflation_as_theft",
        "label": "Inflation / Money Creation as Theft and Redistribution",
        "queries": [
            "inflation redistributes real income illegitimate gains",
            "money production redistributes wealth from poor to rich",
            "redistribution unauthorized taking of property",
        ],
    },
    {
        "id": "voluntariness_principle",
        "label": "The Voluntariness Principle",
        "queries": [
            "voluntary cooperation without violating property rights",
            "giving must not be reluctantly or under compulsion",
            "free responsible initiatives private individuals",
        ],
    },
    {
        "id": "christian_moral_tradition",
        "label": "The Scholastic / Christian Moral Tradition",
        "queries": [
            "scholastic tradition Aquinas Oresme natural law moral reasoning",
            "Scripture Eighth Commandment Romans 13 biblical ethics",
            "Christian morals economics compatible Austrian",
        ],
    },
    {
        "id": "government_as_beneficiary",
        "label": "Government as Beneficiary of the Unjust System",
        "queries": [
            "government main beneficiary inflation unjust system",
            "state third party no ownership claim inserts itself",
            "legal monopolies instruments of social injustice",
        ],
    },
    {
        "id": "institutional_spheres_of_authority",
        "label": "Institutional Design — Proper Spheres of Authority",
        "queries": [
            "poverty relief assigned to individuals church not state",
            "government should not run banks or produce paper money",
            "institutional design coercive mechanism incompatible voluntary",
        ],
    },
    {
        "id": "debasement_as_theft",
        "label": "Debasement as the Historical Paradigm of Government Theft",
        "queries": [
            "debasement standard form of inflation altering coins",
            "debasement inherently unjust never permissible",
            "Mosaic theocracy authorized compulsory redistribution",
        ],
    },
]


async def find_matching_passages(
    container,
    query_text: str,
    target_doc_id: UUID,
    k: int = 5,
    min_score: float = 0.0,
):
    """Search for passages in target_doc matching query_text."""
    # Get all passage IDs for the target document
    all_passages = await container.passages.get_by_document(target_doc_id)
    candidate_ids = [p.id for p in all_passages]

    if not candidate_ids:
        return []

    # Do vector search against those candidates
    query_vec = await container.embedding.embed(query_text)
    hits = await container.passages.vector_search(
        query_vec,
        container.embedding.model_name,
        container.embedding.model_version,
        candidate_ids,
        k,
    )

    return [(pid, score) for pid, score in hits if score >= min_score]


async def main():
    settings = load_settings()
    container = await build_container(settings)

    try:
        # Load all passages for both documents
        article_passages = await container.passages.get_by_document(ARTICLE_DOC_ID)
        book_passages = await container.passages.get_by_document(BOOK_DOC_ID)

        print(f"Article passages: {len(article_passages)}")
        print(f"Book passages: {len(book_passages)}")
        print()

        edge_count = 0

        for theme in THEMES:
            print(f"=== {theme['label']} ===")

            # For each theme, find the best matching passages in each document
            article_hits = []
            book_hits = []

            for query in theme["queries"]:
                a_hits = await find_matching_passages(
                    container, query, ARTICLE_DOC_ID, k=3, min_score=0.3
                )
                b_hits = await find_matching_passages(
                    container, query, BOOK_DOC_ID, k=3, min_score=0.3
                )
                article_hits.extend(a_hits)
                book_hits.extend(b_hits)

            # Deduplicate by passage ID, keeping highest score
            article_best: dict[UUID, float] = {}
            for pid, score in article_hits:
                if pid not in article_best or score > article_best[pid]:
                    article_best[pid] = score

            book_best: dict[UUID, float] = {}
            for pid, score in book_hits:
                if pid not in book_best or score > book_best[pid]:
                    book_best[pid] = score

            # Take top passages from each
            top_article = sorted(article_best.items(), key=lambda x: x[1], reverse=True)[:3]
            top_book = sorted(book_best.items(), key=lambda x: x[1], reverse=True)[:3]

            if not top_article or not top_book:
                print(f"  (no matches found)")
                print()
                continue

            print(f"  Article matches: {len(top_article)}")
            for pid, score in top_article:
                passage = await container.passages.get(pid)
                preview = passage.text[:100].replace("\n", " ") if passage else "?"
                print(f"    [{score:.3f}] {preview}...")

            print(f"  Book matches: {len(top_book)}")
            for pid, score in top_book:
                passage = await container.passages.get(pid)
                preview = passage.text[:100].replace("\n", " ") if passage else "?"
                print(f"    [{score:.3f}] {preview}...")

            # Create edges between the best article passages and best book passages
            async with transaction(container.engine) as tx:
                for a_pid, a_score in top_article:
                    for b_pid, b_score in top_book:
                        combined_confidence = min(a_score, b_score)
                        # Only create edges where both sides have reasonable relevance
                        if combined_confidence < 0.35:
                            continue

                        draft = EdgeDraft(
                            source_kind=NodeKind.passage,
                            source_id=a_pid,
                            target_kind=NodeKind.passage,
                            target_id=b_pid,
                            relation_type="conceptual_overlap",
                            attributes={
                                "theme": theme["id"],
                                "theme_label": theme["label"],
                                "source_document": "article",
                                "target_document": "book",
                                "source_score": round(a_score, 4),
                                "target_score": round(b_score, 4),
                            },
                            confidence=round(combined_confidence, 4),
                        )
                        edge = await container.edges.insert(tx, draft)
                        edge_count += 1

            print()

        # Also create a document-level edge
        async with transaction(container.engine) as tx:
            doc_edge = EdgeDraft(
                source_kind=NodeKind.document,
                source_id=ARTICLE_DOC_ID,
                target_kind=NodeKind.document,
                target_id=BOOK_DOC_ID,
                relation_type="conceptual_overlap",
                attributes={
                    "themes": [t["id"] for t in THEMES],
                    "description": (
                        "Both works argue that redistribution through government "
                        "coercion is a violation of property rights regardless of "
                        "good intentions, and that the proper alternative is "
                        "voluntary action by individuals and private institutions."
                    ),
                    "article_framework": "biblical_protestant",
                    "book_framework": "scholastic_catholic",
                    "article_domain": "fiscal_policy",
                    "book_domain": "monetary_policy",
                },
                confidence=0.95,
            )
            await container.edges.insert(tx, doc_edge)
            edge_count += 1

        print(f"Created {edge_count} edges total.")

    finally:
        await container.close()


if __name__ == "__main__":
    asyncio.run(main())

File: crates/marginalia-parse/src/bin/discover_overlaps.rs
//! `scripts/discover_overlaps.py` (pure parts): overlap edge assembly.
//!
//! The hybrid search and the database writes stay Python; what ports is the
//! ranking both sides share: the score threshold, dedup-keeping-max, the
//! top-three ordering, the combined-confidence gate, and the edge drafts —
//! including the document-level edge. `--themes` dumps the eight analysis
//! themes; `--process` ranks caller-supplied hits for one theme.

use std::collections::HashMap;

use serde_json::{json, Value};

/// The eight conceptual themes the analysis searches from both sides.
fn themes() -> Value {
    json!([
        {"id": "property_rights_as_moral_foundation",
         "label": "Property Rights as a Moral Foundation",
         "queries": [
            "property rights moral foundation sacred inviolable justice",
            "taking property without authorization is theft",
            "private property sacred commutative justice"]},
        {"id": "government_coercion",
         "label": "Government Force / Coercion as the Mechanism of the State",
         "queries": [
            "government monopoly on legitimate use of force compel compliance",
            "compulsion coercion by the ruler government force",
            "forcing citizens to use money of the government's choice"]},
        {"id": "inflation_as_theft",
         "label": "Inflation / Money Creation as Theft and Redistribution",
         "queries": [
            "inflation redistributes real income illegitimate gains",
            "money production redistributes wealth from poor to rich",
            "redistribution unauthorized taking of property"]},
        {"id": "voluntariness_principle",
         "label": "The Voluntariness Principle",
         "queries": [
            "voluntary cooperation without violating property rights",
            "giving must not be reluctantly or under compulsion",
            "free responsible initiatives private individuals"]},
        {"id": "christian_moral_tradition",
         "label": "The Scholastic / Christian Moral Tradition",
         "queries": [
            "scholastic tradition Aquinas Oresme natural law moral reasoning",
            "Scripture Eighth Commandment Romans 13 biblical ethics",
            "Christian morals economics compatible Austrian"]},
        {"id": "government_as_beneficiary",
         "label": "Government as Beneficiary of the Unjust System",
         "queries": [
            "government main beneficiary inflation unjust system",
            "state third party no ownership claim inserts itself",
            "legal monopolies instruments of social injustice"]},
        {"id": "institutional_spheres_of_authority",
         "label": "Institutional Design — Proper Spheres of Authority",
         "queries": [
            "poverty relief assigned to individuals church not state",
            "government should not run banks or produce paper money",
            "institutional design coercive mechanism incompatible voluntary"]},
        {"id": "debasement_as_theft",
         "label": "Debasement as the Historical Paradigm of Government Theft",
         "queries": [
            "debasement standard form of inflation altering coins",
            "debasement inherently unjust never permissible",
            "Mosaic theocracy authorized compulsory redistribution"]}
    ])
}

/// Python's `round(x, 4)`: half to even on the exact binary value.
///
/// Scaling first and rounding the product double-rounds: `1.12345 * 10000`
/// is exactly `11234.5` in `f64`, which ties to even (`1.1234`), while the
/// exact binary value sits above the tie (`1.1235`). Rounding the correctly
/// rounded twenty-place expansion instead — which settles every tie a
/// score this size can hold — matches `round()` exactly.
fn round4(x: f64) -> f64 {
    if !x.is_finite() || x == 0.0 {
        return x;
    }
    let neg = x < 0.0;
    let rendered = format!("{:.20}", x.abs());
    let (int, frac) = rendered.split_once('.').unwrap_or((rendered.as_str(), ""));
    let mut digits: Vec<u8> = int.bytes().chain(frac.bytes().take(4)).collect();
    let mut int_len = int.len();
    let fifth = frac.as_bytes().get(4).copied().unwrap_or(b'0');
    let rest_zero = frac
        .as_bytes()
        .get(5..)
        .is_none_or(|rest| rest.iter().all(|b| *b == b'0'));
    let odd_kept = digits.last().is_some_and(|d| d % 2 == 1);
    if fifth > b'5' || (fifth == b'5' && (!rest_zero || odd_kept)) {
        // Add one unit in the last kept place, carrying left past the
        // point; a carry out front (9.9999 to 10.0000) shifts it right.
        let mut carry = true;
        for digit in digits.iter_mut().rev() {
            if !carry {
                break;
            }
            if *digit == b'9' {
                *digit = b'0';
            } else {
                *digit += 1;
                carry = false;
            }
        }
        if carry {
            digits.insert(0, b'1');
            int_len += 1;
        }
    }
    digits.truncate(int_len + 4);
    let text = format!(
        "{}{}.{}",
        if neg { "-" } else { "" },
        String::from_utf8_lossy(&digits[..int_len.min(digits.len())]),
        String::from_utf8_lossy(&digits[int_len.min(digits.len())..])
    );
    text.parse::<f64>().unwrap_or(x)
}
/// Deduplicate hits keeping the highest score per passage, then the top
/// three by score. The sort is stable, so ties keep first-seen order —
/// exactly the dict-then-sorted order upstream.
fn top_three(hits: &[(String, f64)], min_score: f64) -> Vec<(String, f64)> {
    let mut best: HashMap<&str, f64> = HashMap::new();
    let mut order: Vec<&str> = Vec::new();
    for (pid, score) in hits {
        if *score < min_score {
            continue;
        }
        match best.get(pid.as_str()) {
            Some(current) if *current >= *score => {}
            _ => {
                if !best.contains_key(pid.as_str()) {
                    order.push(pid);
                }
                best.insert(pid, *score);
            }
        }
    }
    let mut ranked: Vec<(String, f64)> = order
        .into_iter()
        .map(|pid| (pid.to_owned(), best[pid]))
        .collect();
    ranked.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    ranked.truncate(3);
    ranked
}

fn process(
    theme_id: &str,
    theme_label: &str,
    article_hits: &[(String, f64)],
    book_hits: &[(String, f64)],
) -> Vec<Value> {
    let top_article = top_three(article_hits, 0.3);
    let top_book = top_three(book_hits, 0.3);
    let mut edges = Vec::new();
    for (a_pid, a_score) in &top_article {
        for (b_pid, b_score) in &top_book {
            let combined = a_score.min(*b_score);
            // Only edges where both sides show reasonable relevance.
            if combined < 0.35 {
                continue;
            }
            edges.push(json!({
                "source_kind": "passage",
                "source_id": a_pid,
                "target_kind": "passage",
                "target_id": b_pid,
                "relation_type": "conceptual_overlap",
                "attributes": {
                    "theme": theme_id,
                    "theme_label": theme_label,
                    "source_document": "article",
                    "target_document": "book",
                    "source_score": round4(*a_score),
                    "target_score": round4(*b_score),
                },
                "confidence": round4(combined),
            }));
        }
    }
    edges
}

fn read_hits(path: &str) -> Vec<(String, f64)> {
    let raw = std::fs::read_to_string(path).unwrap();
    serde_json::from_str::<Vec<Value>>(&raw)
        .unwrap()
        .iter()
        .map(|hit| {
            (
                hit[0].as_str().unwrap().to_owned(),
                hit[1].as_f64().unwrap(),
            )
        })
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.iter().any(|a| a == "--themes") {
        println!("{}", themes());
        return;
    }
    if args.iter().any(|a| a == "--doc-edge") {
        let ids: Vec<Value> = themes()
            .as_array()
            .unwrap()
            .iter()
            .map(|theme| theme["id"].clone())
            .collect();
        println!(
            "{}",
            json!({
                "source_kind": "document",
                "target_kind": "document",
                "relation_type": "conceptual_overlap",
                "attributes": {
                    "themes": ids,
                    "description": "Both works argue that redistribution through government coercion is a violation of property rights regardless of good intentions, and that the proper alternative is voluntary action by individuals and private institutions.",
                    "article_framework": "biblical_protestant",
                    "book_framework": "scholastic_catholic",
                    "article_domain": "fiscal_policy",
                    "book_domain": "monetary_policy",
                },
                "confidence": 0.95,
            })
        );
        return;
    }
    let get = |name: &str| {
        args.iter()
            .position(|a| a == name)
            .and_then(|i| args.get(i + 1).cloned())
            .unwrap_or_else(|| {
                eprintln!("{name} is required");
                std::process::exit(2);
            })
    };
    let edges = process(
        &get("--theme"),
        &get("--label"),
        &read_hits(&get("--article-hits")),
        &read_hits(&get("--book-hits")),
    );
    println!("{}", Value::Array(edges));
}
